include/glibby/spatial/kdTree.h

```diff
@@ -0,0 +1,178 @@
+#pragma once
+
+#include <vector>
+#include <map>
+#include <queue>
+#include "glibby/primitives/point.h"
+#include <iostream>
+#include <string>
+
+namespace glibby
+{
+	template<typename T>
+	struct Node
+	{
+		T val;
+		Node<T>* left;
+		Node<T>* right;
+
+		Node() { left = NULL; right = NULL; }
+		Node(T newVal) { val = newVal; left = NULL; right = NULL; }
+	};
+
+	template<typename T, int N>
+	class kdTree {
+	private:
+		std::vector<glibby::Point<T, N>> points;
+		Node<glibby::Point<T, N>>* parent = NULL;
+
+		bool find(Node<glibby::Point<T, N>>*& node, glibby::Point<T, N> point) {
+			if (!node) {
+				return false;
+			}
+
+			if (node->val == point) {
+				return true;
+			}
+			return find(node->left, point) || find(node->right, point);
+		}
+
+		void construct_tree(std::vector<glibby::Point<T, N>>& points, Node<glibby::Point<T, N>>*& node, int depth) {
+			if (points.size() == 0) {
+				return;
+			}
+
+			if (points.size() == 1) {
+				node = new Node(points[0]);
+				return;
+			}
+
+			std::vector<Point<T, N>> left;
+			std::vector<Point<T, N>> right;
+
+			std::map<int, std::vector<Point<T, N>>> mp;
+			for (int i = 0; i < points.size(); i++) {
+				mp[points[i].coord[depth]].push_back(points[i]);
+			}
+
+			int half = points.size() / 2;
+			int index = 0;
+			for (auto x : mp) {
```

Comment:
make this an auto const&, see https://stackoverflow.com/questions/26281979/c-loop-through-map